### PR TITLE
⏳ Resolve outstanding push request

### LINKED ISSUE #770: 🦦 ResolveOutstandingPushRequest Sync Step
## Description
ResolveOutstandingPushRequest is the [first and fourth step in sync sequence ](https://github.com/toggl/mobileapp/blob/54bb2f1194b97fac59113222caaf390a4acf1581/Toggl.Core/Sync/v2/SyncManagerV2.cs#L191)

See [Xamarin code](https://github.com/toggl/mobileapp/blob/54bb2f1194b97fac59113222caaf390a4acf1581/Toggl.Core/Interactors/Sync/ResolveOutstandingPushRequestInteractor.cs#L31) for reference

## Dependencies
Everything connected to push result processing must be ready 

## Definition of done
- [ ] `OutstandingPush` request in our `SyncApi` and network clients is implemented
- [ ] OutstandingPush request is called
- [ ] OutstandingPush response is processed
- [ ] pushRequestIdentifier is cleared 
- [ ] Add tests if possible

### PR DESCRIPTION
**Pull request by simonrozsival**
### Summary
This is a port of the `ResolveOutstandingPushRequestInteractor` from the Xamarin app.

### Relationships
Closes #770 

### Pull Request checklist
- [x] The code does not hinder the app's performance
- [x] The code does not introduce regressions
- [x] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests
- [ ] ~One of our [live or file custom templates](https://github.com/toggl/android/tree/master/.idea/templates) was used in this PR~

### Review template

- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones

- [ ] The code considers light and dark theme
- [ ] The code considers RTL and does not mention left/right
- [ ] The code considers i18n
  - [ ] There are no hardcoded strings in the UI
  - [ ] All strings are included in every language we support OR there's an issue requesting new translations
- [ ] There are no hard-coded numbers referring to dimensions or font sizes
- [ ] The code considers UI elements that are similar and extracts their similarities into styles to avoid duplication

## Merge permissions
🦑

### CODE REVIEW

File: common-android/src/main/java/com/toggl/common/services/PushRequestIdStorage.kt [hunk 1]
```diff
@@ -21,7 +21,8 @@ class PushRequestIdStorage @Inject constructor(
     fun set(id: UUID?) {
         sharedPreferences.edit { putString(pushRequestIdSharedPreferencesKey, id?.toString() ?: noPushRequestIdDefaultValue) }
     }
+
+    fun clear() = set(null)
+    fun hasOutstandingRequest() = get() != null
 }
```
Comment (heytherewill) [inline common-android/src/main/java/com/toggl/common/services/PushRequestIdStorage.kt]: I don't like this change. This basically means now we have to mock the behavior of `clear` and `hasOutstandingRequest`, when we could much easier have then configured by simply messing with the behavior of `set`

Comment (simonrozsival) [reply common-android/src/main/java/com/toggl/common/services/PushRequestIdStorage.kt]: I actually wanted to do this in order to be able to mock `hasOutstandingRequest`. I didn't like that I had to mock the dependency of `PushRequestIdStorage` in order to test how it is being used. That's one level of nesting too much :-/

I changed all the places where this needs to be mocked, it's was just about changing `set(null)` to `clear()` in `RepositoryTest.kt`. It doesn't require any additional changes.

Are we gonna use `PushRequestIdStorage` anywhere else except for `PushSync`, `ResolveOutstandingPushRequest`, and `Repository`? I don't think we will need it anywhere else 🤔

**Review by heytherewill — APPROVED** [review]
Looks good. I won't fight for the mocking thing because, as you said, it's quite simple, but in general I still prefer extensions for keeping the API sane

**Comment by heytherewill** [comment]
One question, though: Should we not retry only a certain amount of times and then give up? 🤔

File: sync/src/test/java/com/toggl/sync/sequence/ResolveOutstandingPushRequestTest.kt [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.jupiter.api.DisplayName
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertDoesNotThrow
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.MethodSource
+import java.util.UUID
+import java.util.stream.Stream
+
+@DisplayName("The PushSync step")
```
Comment (semanticer) [inline sync/src/test/java/com/toggl/sync/sequence/ResolveOutstandingPushRequestTest.kt:31]: Probably © & 🍝 error?
```suggestion
@DisplayName("The ResolveOutstandingPushRequest step")
```

**Review by semanticer — APPROVED** [review]
Good 👍

File: sync/src/test/java/com/toggl/sync/sequence/ResolveOutstandingPushRequestTest.kt (hunk 2 above)
Comment (simonrozsival) [reply sync/src/test/java/com/toggl/sync/sequence/ResolveOutstandingPushRequestTest.kt:31]: You're right, thanks!

**Comment by simonrozsival** [comment]
@heytherewill Yep, we might do that. Let's finish the initial implementation and add that once we start testing it 👍 (please create an issue)

**Pull request merged**